### PR TITLE
Trigger now if workflow paused

### LINKED ISSUE #5963: Allow trigger-now in a paused workflow?
There is a valid use case (especially since the demise of the old `cylc submit` command for running individual tasks sans scheduler) early on in workflow development and debugging, for starting a workflow paused and then manually triggering individual tasks to run.

Currently "paused" means "don't run anything", so if you trigger a task in a paused workflow, the task won't run until the workflow gets un-paused.

I think we probably intended `cylc hold` for this sort of thing, but that doesn't do it without additional seemingly unnecessary effort: holding an already-active task (even via a "hold all" pattern) does not prevent downstream future tasks from running unless every one of those has also been specifically held.

So ... is there any good reason not to allow manual triggering (in the "run right now" sense) of tasks in a paused workflow?

See cylc/cylc-ui#1651

### LINKED ISSUE #6398: Held tasks need multiple triggers to run?
I am finding that if I have a held workflow (not paused), if I trigger a task it appears to do nothing. But, if I trigger it twice, it will then run the task. See the attached gif.

multiple_trigger_to_run

The scheduler log shows this

```
2024-10-08T00:42:22Z INFO - Command "force_trigger_tasks" received. ID=8b832ccb-2fba-46c7-a389-602fb984d60c
    force_trigger_tasks(flow=['all'], flow_wait=False, tasks=['20241006T1800Z/glu_ops_process_background_iasi'])
2024-10-08T00:42:22Z INFO - Command "force_trigger_tasks" actioned. ID=8b832ccb-2fba-46c7-a389-602fb984d60c
2024-10-08T00:42:22Z INFO - [20241006T1800Z/glu_ops_process_background_iasi:waiting(held)] => waiting(held)(queued)
2024-10-08T00:42:29Z INFO - Command "force_trigger_tasks" received. ID=883ccef9-4b47-4cac-9258-26056328462b
    force_trigger_tasks(flow=['all'], flow_wait=False, tasks=['20241006T1800Z/glu_ops_process_background_iasi'])
2024-10-08T00:42:29Z INFO - Command "force_trigger_tasks" actioned. ID=883ccef9-4b47-4cac-9258-26056328462b
```

If I read that correctly, the trigger pushes it to the `queued` state, and then a second trigger pushes to submit the job. I had expected that triggering would shift it twice.

I'm unsure if this is a Cylc-flow question or a UI question, so I've put it here as that is how I'm looking at it.

### PR DESCRIPTION
**Pull request by hjoliver**
Close #5963
Supesede #6192 (it was getting messy)
Close #6398

------

New behaviour for triggering when the workflow is paused

**DEFAULT: run triggered tasks NOW even if the workflow is paused**
  - supports the common use case *"I want to pause the workflow and just trigger individual tasks"*
  - (NIWA priority raised by both ops and research users in early Cylc 8 migration meetings)

**OPTION: `--on-resume` - run triggered tasks only when the paused workflow is resumed**
  - this supports [the broadcast-to-past-tasks workaround](https://github.com/cylc/cylc-flow/pull/6192#issuecomment-2486785465) until broadcast is fixed (prob. 8.5.0)
  - the option can be removed in the future, once broadcast is fixed, if there are no other use cases
  

-----

Implementation:
- separate start of job submission from queue release, in the scheduler
- manual trigger adds tasks to `tasks_to_trigger_(now|on_resume)` lists, after state reset and queue wrangling
- if these lists are populated, the scheduler will release tasks from them as appropriate
- (queues are not processed if the scheduler is paused)

-------

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] Changelog entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at https://github.com/cylc/cylc-doc/pull/778
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

**Comment by hjoliver** [comment]
(I just need to extend the tests now that I've made trigger-when-paused optional).

**Comment by dwsutherland** [comment]
> (queues are not processed if the scheduler is paused)

Does this mean if I trigger when pause, and a queue limit is reached, that it will enter a queue but not be released when the queue "empties" of running tasks?

Will test.

**Comment by hjoliver** [comment]
> > (queues are not processed if the scheduler is paused)
> 
> Does this mean if I trigger when pause, and a queue limit is reached, that it will enter a queue but not be released when the queue "empties" of running tasks?

Yes.  Pausing the scheduler halts all automatic job submission, so that means queued tasks won't automatically run. But if you want the queued task to run before resuming the workflow you can just trigger it again before resuming.

**Comment by hjoliver** [comment]
(I've just tweak the `pause` and `trigger` command help in an attempt to make that clearer @dwsutherland )

**Review by dwsutherland — APPROVED** [review]
LGTM, reacts as expect with queued/unqueued 👍 
```
    [[queues]]
        [[[THIS]]]
            limit = 1
            members = baz, zip, toz
    [[graph]]
        R1 = "prep => foo"
        PT12H = "foo[-PT12H] => foo => baz & zip & toz => bar"
```
image

File: tests/functional/restart/58-waiting-manual-triggered.t [hunk 1]
```diff
@@ -40,6 +40,7 @@ __EOF__
 
 # It should restart and shut down normally, not stall with 2/foo waiting on 1/foo.
 workflow_run_ok "${TEST_NAME_BASE}-restart" cylc play --no-detach "${WORKFLOW_NAME}"
+
```
[the remark this answers is not in the document]
Comment (hjoliver) [reply tests/functional/restart/58-waiting-manual-triggered.t:43]: (What's funny about that? I'm dying to know!)

[the remark this answers is not in the document]
Comment (hjoliver) [reply tests/functional/restart/58-waiting-manual-triggered.t:43]: Haha, right 😆  I probably had to visit that file somewhere during branch development, and left a trace.

File: cylc/flow/scripts/trigger.py [hunk 2]
```diff
@@ -17,11 +17,17 @@
 
 """cylc trigger [OPTIONS] ARGS
 
-Force tasks to run regardless of prerequisites.
+Manually trigger tasks regardless of prerequisites, even in a paused workflow.
```
Comment (MetRonnie) [inline cylc/flow/scripts/trigger.py:20]: IMO avoid repeating the command verb in the description as that doesn't expand on what it does
```suggestion
Force task(s) to run regardless of prerequisites, even in a paused workflow.
```

File: cylc/flow/scripts/trigger.py [hunk 3]
```diff
@@ -96,7 +104,16 @@ def get_option_parser() -> COP:
         multiworkflow=True,
         argdoc=[FULL_ID_MULTI_ARG_DOC],
     )
+
     add_flow_opts(parser)
+
+    parser.add_option(
+        "--on-resume",
+        help=r"Run triggered tasks once a paused workflow is resumed.",
```
Comment (MetRonnie) [inline cylc/flow/scripts/trigger.py:112]:
```suggestion
        help=(
            "If the workflow is paused, wait until it is resumed before "
            "running the triggered task(s)."
        ),
```

File: cylc/flow/scripts/trigger.py [hunk 4]
```diff
@@ -17,11 +17,17 @@
...
-* Triggering an unqueued waiting task queues it, regardless of prerequisites.
-* Triggering a queued task submits it, regardless of queue limiting.
-* Triggering an active task has no effect (it already triggered).
+Triggering an unqueued task queues it, to run when released by the queue.
+Triggering a queued task runs it immediately regardless of queue limiting.
+So you may need to trigger tasks twice if queue limiting is in effect.
+
+If the workflow is paused queued waiting tasks will not run (unless manually
+triggered) until the workflow is resumed, even if the queue empties out. This
+includes tasks queued by manual triggering, when queue limits are in effect.
```
Comment (MetRonnie) [inline cylc/flow/scripts/trigger.py:28]:
```suggestion
Triggering a queued task runs it immediately.

Triggering a task that is not yet queued (i.e. "queue-limited") will queue it;
it will run when released from the queue. So you may need to trigger tasks
twice if queue limiting is in effect.
```

File: cylc/flow/task_queues/independent.py [hunk 5]
```diff
@@ -40,6 +40,21 @@ def push_task(self, itask: 'TaskProxy') -> None:
         if itask.tdef.name in self.members:
             self.deque.appendleft(itask)
 
+    def push_task_if_limited(
+        self, itask: 'TaskProxy', active: Counter[str]
+    ) -> bool:
+        """Queue task if in my membership and the queue limit is reached."""
+        n_active: int = 0
+        for mem in self.members:
+            n_active += active[mem]
```
Comment (MetRonnie) [inline cylc/flow/task_queues/independent.py:49]:
```suggestion
        n_active = sum(active[mem] for mem in self.members)
```

File: cylc/flow/task_queues/independent.py [hunk 6]
```diff
@@ -113,34 +128,37 @@ def __init__(self,
...
-    def force_release_task(self, itask: 'TaskProxy') -> None:
+    def force_release_task(self, itask: 'TaskProxy') -> bool:
         """Remove a task from whichever queue it belongs to.
 
-        To be returned when release_tasks() is next called.
+        Return True if released, else False.
         """
-        if self.remove_task(itask):
-            self.force_released.add(itask)
+        return self.remove_task(itask)
```
Comment (MetRonnie) [inline cylc/flow/task_queues/independent.py:161]: Not needed anymore?

File: cylc/flow/scripts/pause.py [hunk 7]
```diff
@@ -18,9 +18,9 @@
 
 """cylc pause [OPTIONS] ARGS
 
-Pause a workflow.
+Suspend all automatic job submission until the workflow is resumed.
```
[the remark this answers is not in the document]
Comment (oliver-sanders) [reply cylc/flow/scripts/pause.py:21]: To help users navigate the CLI, try to keep the first line as a quick at-a-glance description and put the remaining detail on following lines.

Suggest:

```suggestion
Suspend automatic job submission.
```

Or:

```suggestion
Pause a workflow.
```

With the details of *exactly* what this means in the following paragraph.

**Comment by oliver-sanders** [comment]
@hjoliver, by bypassing queue processing for manual trigger, this PR may close https://github.com/cylc/cylc-flow/issues/6398

File: tests/integration/conftest.py [hunk 8]
```diff
@@ -420,6 +421,8 @@ def _disable_submission(schd: 'Scheduler') -> 'Set[TaskProxy]':
 
         def _submit_task_jobs(_, itasks, *args, **kwargs):
             nonlocal submitted_tasks
+            for itask in itasks:
+                itask.state_reset(TASK_STATUS_SUBMITTED)
```
Comment (oliver-sanders) [inline tests/integration/conftest.py:425]: FYI: This fixture is intended to capture submission attempts, it wasn't intended to mock the submission process/outcome.

File: cylc/flow/scripts/trigger.py (hunk 3 above)
Comment (MetRonnie) [reply cylc/flow/scripts/trigger.py:112]:
```suggestion
        help=(
            "If the workflow is paused, wait until it is resumed before "
            "running the triggered task(s). DEPRECATED - this may be "
            "removed at Cylc 8.5."
        ),
```

Comment (oliver-sanders) [reply cylc/flow/scripts/trigger.py:112]: (can probably replace the "may" with "will")

Comment (MetRonnie) [reply cylc/flow/scripts/trigger.py:112]: I'm wary of making promises (how many "final" Cylc 7 release are we at?) 😅

File: tests/integration/test_flow_assignment.py [hunk 9]
```diff
@@ -110,53 +113,65 @@ async def test_flow_assignment(
...
         # By default inactive tasks get all active flows.
         do_command(['1/a'], flow=[])
         assert schd.pool._get_task_by_id('1/a').flow_nums == {1, 2, 3}
 
+        # set --pre=all should not cause a task to submit in a paused workflow
+        assert len(submitted_tasks) == 0
+        # but triggering it should
+        schd.pool.force_trigger_tasks(['1/a'], [])
+        schd.release_tasks_to_run()
+        assert len(submitted_tasks) == 1
```
Comment (MetRonnie) [inline tests/integration/test_flow_assignment.py:173]: Please could you put this new bit in a separate test and revert the splitting of `trigger` and `set` from this test; a key feature of this test was that is checks equity between `trigger` and `set` for the same flow nums

File: tests/integration/test_force_trigger.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+    The usual queue limiting behaviour is expected.
+
+    https://github.com/cylc/cylc-flow/issues/6192
+
+    """
+    id_ = flow({
+        'scheduler': {
+            'allow implicit tasks': True,
+        },
```
Comment (MetRonnie) [inline tests/integration/test_force_trigger.py:42]: (This is automatically applied in integration tests)
```suggestion
```

File: tests/integration/test_scheduler.py [hunk 11]
```diff
@@ -346,6 +346,17 @@ async def test_restart_timeout(
 
     See: https://github.com/cylc/cylc-flow/issues/5078
     """
+    one_conf.update(
+        {
+            'runtime': {
+                'one': {
+                    'simulation': {
+                        'default run length': 'PT5S'
```
Comment (MetRonnie) [inline tests/integration/test_scheduler.py:354]: I've got an auxiliary PR to simplify this back down: https://github.com/hjoliver/cylc-flow/pull/60

File: cylc/flow/scripts/trigger.py (hunk 3 above)
Comment (oliver-sanders) [reply cylc/flow/scripts/trigger.py:112]: Understood, but we can amend that in the following release in the, hopefully, unlikely event that we need to delay it.

File: cylc/flow/network/schema.py [hunk 12]
```diff
@@ -2180,20 +2180,24 @@ class Arguments(TaskMutation.Arguments, FlowMutationArguments):
...
 
-            Warning: waiting tasks that are queue-limited will be queued if
-            triggered, to submit as normal when released by the queue; queued
-            tasks will submit immediately if triggered, even if that violates
-            the queue limit (so you may need to trigger a queue-limited task
-            twice to get it to submit immediately).
+            Triggering an unqueued task queues it, to run when queue-released.
+            Triggering a queued task runs it now regardless of queue limiting.
+
+            The "on resume" option waits for a paused workflow to be resumed.
```
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/network/schema.py:2188]: What exactly don't you like about the original? 

Well, "a task that is not yet queued" is technically not "queue-limited".  We, or at least I, have always taken "queue-limited" to mean what happens when the queue's active task limit has been reached, i.e. when a task is ready to run but the queue is holding it back. 

I was trying to avoid making that distinction, as something potentially requiring user thought, by simply saying that triggering a task queues it, to run when released by the queue - which is immediately if queue-limiting is not in effect, or later if it is in effect.

File: cylc/flow/network/schema.py [hunk 13]
```diff
@@ -2180,20 +2180,24 @@ class Arguments(TaskMutation.Arguments, FlowMutationArguments):
...
             Valid for: paused, running workflows.
         ''')
         resolver = partial(mutator, command='force_trigger_tasks')
 
     class Arguments(TaskMutation.Arguments, FlowMutationArguments):
-        ...
+        on_resume = Boolean(
+            default_value=False,
+            description=sstrip('''
+                Run triggered tasks once the paused workflow is resumed.
```
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/network/schema.py:2198]: Fair enough, perhaps sometimes I'm over-zealous in trying to be concise.

File: cylc/flow/scripts/pause.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/scripts/pause.py:21]: > Grammatically I think it's either "suspend all automatic job submissions" or "suspend automatic job submission".

Agreed, good call.

However, I'll take @oliver-sanders first suggestion.  The second one essentially just repeats the command name, which seems a bit pointless.

File: cylc/flow/scripts/trigger.py (hunk 3 above)
Comment (hjoliver) [reply cylc/flow/scripts/trigger.py:112]:
```suggestion
        help=(
            "If the workflow is paused, wait until it is resumed before "
            "running the triggered task(s). DEPRECATED - this will be "
            "removed at Cylc 8.5."
        ),
```

Comment (hjoliver) [reply cylc/flow/scripts/trigger.py:112]: OK going with "will".

File: cylc/flow/scripts/pause.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/scripts/pause.py:21]:
```suggestion
Suspend automatic job submission.
```

File: tests/integration/conftest.py (hunk 8 above)
Comment (hjoliver) [reply tests/integration/conftest.py:425]: I forget why I did that, checking back...

File: cylc/flow/network/schema.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/network/schema.py:2188]: How about this:
```suggestion
            Triggering a task that is not yet queued, queues it.
               
            Triggering a queued task runs it immediately.
                        
            Queues release tasks to run when the count of active tasks
            drops below the queue limit. So, depending on activity, you
            may need to trigger a task twice to make it run immediately. 
```

File: cylc/flow/scripts/trigger.py (hunk 4 above)
Comment (hjoliver) [reply cylc/flow/scripts/trigger.py:28]: As for the CLI help:
```suggestion
Triggering a task that is not yet queued, queues it.

Triggering a queued task runs it immediately.
                        
Queues release tasks to run when their active task count drops below the queue
limit. So, depending on activity, you may need to trigger a task twice to make
it run immediately. 
```

File: cylc/flow/task_queues/independent.py (hunk 6 above)
Comment (hjoliver) [reply cylc/flow/task_queues/independent.py:159]: True, I've removed the method.

**Comment by hjoliver** [comment]
@oliver-sanders -

> @hjoliver, by bypassing queue processing for manual trigger, this PR may close #6398

That issue is confusing. I think the "bug" initially reported by Tom is just the expected double-trigger requirement to get past the queues. In his description, despite the tasks being held, they did run on the second trigger attempt - but he had expected that manual trigger would skip queues to run immediately after just one attempt.

So I presume you're referring to the presumably-actual-bug that you commented on part way through the discussion: https://github.com/cylc/cylc-flow/issues/6398#issuecomment-2407490625

File: tests/integration/test_flow_assignment.py (hunk 9 above)
Comment (hjoliver) [reply tests/integration/test_flow_assignment.py:173]: The reason I split that into separate tests is that this PR changes that equivalence. 

Trigger and set (in a paused workflow) are more similar on master, because a triggered task does not become "active" until the workflow is resumed, so it can be triggered multiple times to absorb multiple flow numbers. On this branch, a triggered task becomes active immediately and so subsequent triggers get ignored.

In light of that, do you still want me to revert that? I suppose we could have a combined test for what is common, and separate tests for the behavior that differs.

**Comment by oliver-sanders** [comment]
> I think the "bug" initially reported by Tom is just the expected double-trigger requirement to get past a queue.

No, there is definitely a bug lurking here. We have seen situations where tasks will not run no matter how many times you trigger them, until unheld.

Please see https://github.com/cylc/cylc-flow/issues/6398#issuecomment-2407490625 where I show how an `is_held` check in the queue logic could cause a held task to enter a state where it cannot be run without being unheld.

File: cylc/flow/network/schema.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (MetRonnie) [reply cylc/flow/network/schema.py:2188]: Ok, small amendment to your suggestion:
```suggestion
            Triggering a task that is not yet queued will queue it.
               
            Triggering a queued task runs it immediately.
                        
            Queues release tasks to run when their active task count
            drops below the queue limit. So, depending on the task count, you
            may need to trigger a task twice to make it run immediately. 
```

File: tests/integration/test_flow_assignment.py (hunk 9 above)
Comment (MetRonnie) [reply tests/integration/test_flow_assignment.py:173]: > In light of that, do you still want me to revert that? I suppose we could have a combined test for what is common, and separate tests for the behavior that differs.

👆 That

File: cylc/flow/scheduler.py [hunk 14]
```diff
@@ -1703,7 +1740,7 @@ async def _main_loop(self) -> None:
                 self.broadcast_mgr.check_ext_triggers(
                     itask, self.ext_trigger_queue)
 
-            if itask.is_ready_to_run():
+            if itask.is_ready_to_run() and not itask.is_manual_submit:
```
Comment (MetRonnie) [inline cylc/flow/scheduler.py:1743]: What is the reason for this change?

File: cylc/flow/scripts/trigger.py (hunk 4 above)
Comment (hjoliver) [reply cylc/flow/scripts/trigger.py:28]: @MetRonnie - I've tweaked this for consistency with your suggestion for the CLI help, so will merge it.

**Comment by hjoliver** [comment]
(I ran out of time today, but I just need to tweak the tests as requested by @MetRonnie ... )

**Comment by hjoliver** [comment]
> So it looks like this closes https://github.com/cylc/cylc-flow/issues/6398 (congrats on fixing a bug you didn't think existed!).

Great, well done me. I'll add a close directive to the description above...

**Comment by oliver-sanders** [comment]
> Great, well done me. I'll add a close directive to the description above...

(will see if I can get a test for this)

File: cylc/flow/commands.py [hunk 15]
```diff
@@ -448,9 +448,12 @@ async def force_trigger_tasks(
     flow: List[str],
     flow_wait: bool = False,
     flow_descr: Optional[str] = None,
+    on_resume: bool = False
 ):
     """Manual task trigger."""
     validate.is_tasks(tasks)
     validate.flow_opts(flow, flow_wait)
```
Comment (MetRonnie) [inline cylc/flow/commands.py:455]:
```suggestion
    validate.flow_opts(flow, flow_wait)
    if on_resume:
        LOG.warning(
            "The --on-resume option is deprecated and will be removed "
            "at Cylc 8.5."
        )
```

File: cylc/flow/scripts/trigger.py [hunk 16]
```diff
@@ -17,11 +17,17 @@
...
-* Triggering an unqueued waiting task queues it, regardless of prerequisites.
-* Triggering a queued task submits it, regardless of queue limiting.
-* Triggering an active task has no effect (it already triggered).
+Triggering a task that is not yet queued will queue it.
+
+Triggering a queued task runs it immediately.
+
+Queues release tasks to run when their active task count drops below the queue
+limit. So, depending on the task count, you may need to trigger a task twice to make
+it run immediately.
```
Comment (MetRonnie) [inline cylc/flow/scripts/trigger.py:28]:
```suggestion
limit. So, depending on the task count, you may need to trigger a task twice
to make it run immediately.
```

File: cylc/flow/task_pool.py [hunk 17]
```diff
@@ -2109,48 +2119,91 @@ def _get_flow_nums(
...
-        # TODO is this flag still needed, and consistent with "cylc set"?
+    def _force_trigger(self, itask: 'TaskProxy', on_resume: bool = False):
+        """Process a manually triggered task, ready for job submission.
+
+        Assumes the task is in the pool.
+
+        Triggering a queued task will:
+          - run it, regardless of queue limiting
+
+        Triggering an unqueued task will:
```
Comment (MetRonnie) [inline cylc/flow/task_pool.py:2130]: (To me, "unqueued" implies that it was once queued and no longer is)
```suggestion
        Triggering an non-queued task will:
```

File: tests/functional/cylc-trigger/07-kill-trigger/reference.log [hunk 18]
```diff
@@ -0,0 +1,2 @@
+1/a -triggered off [] in flow 1
+1/a -triggered off [] in flow 1
```
Comment (MetRonnie) [inline tests/functional/cylc-trigger/07-kill-trigger/reference.log:2]: The test is not doing a reftest though?

File: cylc/flow/task_pool.py [hunk 19]
```diff
@@ -2109,48 +2119,91 @@ def _get_flow_nums(
...
+                with suppress(KeyError):
+                    # In case previously triggered without --on-resume.
+                    # (It should have run already, but just in case).
+                    self.tasks_to_trigger_now.remove(itask)
+                self.tasks_to_trigger_on_resume.add(itask)
+            else:
+                with suppress(KeyError):
+                    # In case previously triggered with --on-resume.
+                    self.tasks_to_trigger_on_resume.remove(itask)
+                self.tasks_to_trigger_now.add(itask)
```
Comment (MetRonnie) [inline cylc/flow/task_pool.py:2183]:
```suggestion
                self.tasks_to_trigger_on_resume.add(itask)
                # In case previously triggered without --on-resume.
                # (It should have run already, but just in case).
                self.tasks_to_trigger_now.discard(itask)
            else:
                self.tasks_to_trigger_now.add(itask)
                # In case previously triggered with --on-resume.
                self.tasks_to_trigger_on_resume.discard(itask)
```

**Review by MetRonnie — COMMENTED** [review]
Consider this approved pending addressing of comments (including https://github.com/cylc/cylc-flow/pull/6499#discussion_r1869261786)

File: tests/functional/cylc-trigger/07-kill-trigger/reference.log (hunk 18 above)
Comment (oliver-sanders) [reply tests/functional/cylc-trigger/07-kill-trigger/reference.log:2]: Dammit, my fault, I forgot to turn this on in the functional test.

File: cylc/flow/commands.py (hunk 15 above)
Comment (oliver-sanders) [reply cylc/flow/commands.py:455]: What are the bets this option is never used?!

File: cylc/flow/network/schema.py [hunk 20]
```diff
@@ -2180,20 +2180,28 @@ class Arguments(TaskMutation.Arguments, FlowMutationArguments):
...
-            tasks will submit immediately if triggered, even if that violates
-            the queue limit (so you may need to trigger a queue-limited task
-            twice to get it to submit immediately).
+            Triggering a task that is not yet queued will queue it.
+               
+            Triggering a queued task runs it immediately.
+                        
+            Queues release tasks to run when their active task count
+            drops below the queue limit. So, depending on the task count, you
+            may need to trigger a task twice to make it run immediately.
```
Comment (oliver-sanders) [inline cylc/flow/network/schema.py:2191]: > when their active task count drops below the queue limit

Accurate, but maybe not the clearest for users who may not know what Cylc queues are.

Suggestion:
* Added an explanation of what queues are and referred to them as "Cylc" queues to differentiate from job runner queues.
* Updated "depending on the task count" to "on the **active** task count".

```suggestion
            Cylc queues restrict the number of jobs that can submit
            / run simultaneously.
            Queues release tasks to run when their active task count
            drops below the queue limit. So, depending on the active task count, you
            may need to trigger a task twice to make it run immediately. 
```

**Comment by oliver-sanders** [comment]
(have manually run the auto-restart tests on this branch, they all pass)

File: cylc/flow/scheduler.py [hunk 21]
```diff
@@ -642,6 +642,19 @@ async def run_scheduler(self) -> None:
...
+                    for itask in self.pool.get_tasks():
+                        if (
+                            itask.is_manual_submit
+                            and itask.state(TASK_STATUS_WAITING)
+                        ):
+                            itask.waiting_on_job_prep = True
+                            pre_prep_tasks.append(itask)
+
+                    self.start_job_submission(pre_prep_tasks)
+
```
Comment (oliver-sanders) [inline cylc/flow/scheduler.py:657]: I don't think it's right for tasks triggered with `cylc trigger --on-resume` to be run instantly if a workflow is restarted with `cylc play --pause`, however, this option won't last long and may only be used by a couple of people so we can handle some quirks.

File: cylc/flow/scheduler.py [hunk 22]
```diff
@@ -1419,13 +1433,24 @@ def release_queued_tasks(self) -> bool:
...
+                pre_prep_tasks.update(self.pool.tasks_to_trigger_now)
+                self.pool.tasks_to_trigger_now = set()
+
+            if not self.is_paused:
+                # release queued tasks
+                pre_prep_tasks.update(self.pool.release_queued_tasks())
+                if self.pool.tasks_to_trigger_on_resume:
+                    # and manually triggered tasks to run once workflow resumed
+                    pre_prep_tasks.update(self.pool.tasks_to_trigger_on_resume)
+                    self.pool.tasks_to_trigger_on_resume = set()
```
Comment (oliver-sanders) [inline cylc/flow/scheduler.py:1453]: I think codecov is right, these lines are untested.

File: cylc/flow/commands.py (hunk 15 above)
Comment (hjoliver) [reply cylc/flow/commands.py:455]: The point is, BOM need the option to be available *in case it is needed* to address a real problem that they have already encountered in the wild. (Broadcasting to past task that needs to be retriggered).  I would not have been proposing this option if we didn't have that curly one to deal with.

File: cylc/flow/task_pool.py (hunk 19 above)
Comment (hjoliver) [reply cylc/flow/task_pool.py:2183]: Nice, thanks.

File: cylc/flow/scheduler.py (hunk 21 above)
Comment (hjoliver) [reply cylc/flow/scheduler.py:657]: Yeah I was being a bit lazy there. I'll consider fixing that, just in case...

File: cylc/flow/scheduler.py (hunk 22 above)
Comment (hjoliver) [reply cylc/flow/scheduler.py:1453]: Correct, I was waiting for agreement to add the option before adding tests for it, then failed to come back to the PR till now.

File: cylc/flow/network/schema.py (hunk 20 above)
Comment (hjoliver) [reply cylc/flow/network/schema.py:2191]: I like that, will accept and copy...

File: tests/functional/cylc-trigger/07-kill-trigger/reference.log (hunk 18 above)
Comment (hjoliver) [reply tests/functional/cylc-trigger/07-kill-trigger/reference.log:2]: FYI I had to tweak your func test slightly to make it work as a reference test:
```diff
 [scheduler]
     [[events]]
         inactivity timeout = PT1M
+        expected task failures = "1/a"

 [scheduling]
     [[graph]]
@@ -14,6 +15,8 @@
                 # then kill the job
                 cylc__job__poll_grep_workflow_log -E '1/a.*running'
                 cylc kill "${CYLC_WORKFLOW_ID}//1/a"
+                # do not succeed immediately after issuing the kill command or the
+                # workflow may shut down as complete before registering task failure
+                # (This polling grep will never complete, but you know what I mean.)
+                cylc__job__poll_grep_workflow_log -E '1/a.*failed'
             fi
         """
```

**Review by oliver-sanders — APPROVED** [review]
Approved pending resolution of Ronnie's comment above and, ideally, some kind of test for the `--on-resume` option.

File: tests/integration/test_flow_assignment.py (hunk 9 above)
Comment (hjoliver) [reply tests/integration/test_flow_assignment.py]: Done

**Comment by hjoliver** [comment]
Reverted the set/trigger flow-assignment test
Added tests for the new option
Tests all passing
No other changes since approvals ... merging.
